lib/semantics/resolve-names.cpp

```diff
@@ -6051,6 +6062,53 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPLoopConstruct &x) {
   return true;
 }
 
+const parser::Name &OmpAttributeVisitor::GetLoopIndex(
+    const parser::DoConstruct &x) {
+  auto &loopControl{x.GetLoopControl().value()};
+  using Bounds = parser::LoopControl::Bounds;
+  const Bounds &bounds{std::get<Bounds>(loopControl.u)};
+  return bounds.name.thing;
+}
+
+void OmpAttributeVisitor::ResolveSeqLoopIndexInParallelOrTaskConstruct(
+    const parser::Name &iv) {
+  Symbol *symbol{nullptr};
+  std::vector<OmpContext>::reverse_iterator targetIt;
+  for (auto it{ompContext_.rbegin()}; it != ompContext_.rend(); ++it) {
+    if (parallelSet.test(it->directive) ||
+        taskGeneratingSet.test(it->directive)) {
+      symbol = ResolveOmp(iv, Symbol::Flag::OmpPrivate, it->scope);
+      targetIt = it;
+      break;
+    }
+  }
+  if (symbol) {
+    targetIt++;
+    symbol->set(Symbol::Flag::OmpPreDetermined);
+    iv.symbol = symbol;  // adjust the symbol within region
+    for (auto it{ompContext_.rbegin()}; it != targetIt; ++it) {
+      AddToContextObjectWithDSA(*symbol, Symbol::Flag::OmpPrivate, *it);
+    }
+  }
+}
+
+// 2.15.1.1 Data-sharing Attribute Rules - Predetermined
+//   - A loop iteration variable for a sequential loop in a parallel
+//     or task generating construct is private in the innermost such
+//     construct that encloses the loop
+bool OmpAttributeVisitor::Pre(const parser::DoConstruct &x) {
```

Comment:
How the "sequential" aspect is determined here ? Should it be tested that the `parser::DoConstruct` belongs to a `parser::OpenMPLoopConstruct` ?

Reply:
No. The `sequential` here means that do-loop is within OpenMP construct but not associated with `OpenMPLoopConstruct`. Like:
```
!$omp parallel do
do i = 1, 10   <-- associated
   do j = 1, 10   <-- sequential
    ...
   end do
enddo
```

Reply:
How do you make the difference between loops that are associated to `OpenMPLoopConstruct` or not here ?

Reply:
Through line 6102 below, if the loop index is already determined as `OmpPredetermined`, then ignore.